dev/TreeView/TreeView.cpp

```diff
@@ -381,6 +386,7 @@ void TreeView::OnApplyTemplate()
         m_containerContentChangingRevoker = listControl.ContainerContentChanging(winrt::auto_revoke, { this, &TreeView::OnContainerContentChanging });
         m_dragItemsStartingRevoker = listControl.DragItemsStarting(winrt::auto_revoke, { this, &TreeView::OnListControlDragItemsStarting });
         m_dragItemsCompletedRevoker = listControl.DragItemsCompleted(winrt::auto_revoke, { this, &TreeView::OnListControlDragItemsCompleted });
+        m_selectionChangedRevoker = listControl.SelectionChanged(winrt::auto_revoke, { this, &TreeView::OnListControlSelectionChanged });
```

Comment:
@kaiguo Is there a difference between usage for single/multiple selection modes ? Would just forwarding listview's selection changed event work in all cases  ?

Reply:
Yeah TreeView is handling the selection on its own in multi-select mode, so ListView probably won't receive SelectionChanged event in multi-selection. Try monitoring [m_selectedNodes](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/TreeView/ViewModel.h#L78) instead.

Reply:
Interesting, this is getting more complex than I thought :-)

I had a look at `m_selectedNodes` that we could monitor with a call to `VectorChanged`. However, we would get 2 calls to the callback when a single selection change happens (for example selection change from item A to item B: one item is removed, another one is added). I think this would make it hard to send a single `SelectionChanged` event.

A more complex option would be introduce `BeginSelectionChange`/`EndSelectionChange` to start tracking changes to the vector before doing any changes and stop tracking changes after, and then raise the new `SelectionChanged` event.

It could look like this:

1. In `ViewModel.cpp`, add a `BeginSelectionChange` and a `EndSelectionChange` methods
2. `BeginSelectionChange` would subscribe to the changes of the `m_selectedNodes` and initialize 2 collections to track added and removed items.
3. `EndSelectionChange` would unsubscribe to changes the `m_selectedNodes` and raise the `SelectionChanged` event based on the 2 collections filled while `m_selectedNodes` has changed
4. We'd need to change all calls to `UpdateNodeSelection` to wrap them with a pair of `BegingSelectionChange`/`EndSelectionChange`

What do you think? Maybe there is a much simpler alternative I'm not seeing.

Reply:
We definitely need to avoid raising the event twice, as it is confusing for the developers. Correct me if I am wrong but is your idea something like the following:

SelectedNodes changed for the first time -> BeginSelectionChange gets called, creates list, unsubscribes itsself, subscribes EndSelectionChange

SelectedNodes changed for the "second time" -> EndSelectionChange gets called, raises the TreeViews SelectionChanged event with the 2 lists created by BeginSelectionChange, unsubscribes itsself and subscribes BeginSelectionChange

I think we might be able to solve that with a single function and flags indicating wether we were raised twice or not, but I am not quite sure whether we might run into concurrency problems.

Reply:
It's close, but not exactly that.

I don't think we can rely on the fact that `m_selectedNodes` will changed 2 times. You could have one item selected and clear the selection and you'd see only 1 change in the collection (the item being removed).

What I'm proposing instead of to look at the places where we're doing multiple changes to the `m_selectedNodes` vector, and wrap those with a call to `BeginSelectionChange` and `EndSelectionChange`.

Reply:
I think rather than making methods to start/end selection you should use something like this:

        auto clearSelecting = gsl::finally([this]()
            {
                m_currentlySelecting = false;
            });
        m_currentlySelecting = true;

you can see the pattern in use in the radioButtons control, and a few other places in the repo

Reply:
When the selection of the treeview changes, who gets notified first? Is the ListView is the first to notice that? In that case I am not quite sure if we can use this scoping to notice if we got run multiple times.

Reply:
@chingucoding When the TreeView is in multi-selection mode, the TreeView handles the selection and the underlying ListView (m_listControl = TreeViewList) is set to selection mode 'none' in this case. So the treeview should get notified first (and probaby only):

```cs
case winrt::TreeViewSelectionMode::Multiple:
{
      m_listControl.get().SelectionMode(winrt::ListViewSelectionMode::None);
      UpdateItemsSelectionMode(true);
}
```

In single-selection mode, however, I believe TreeView relies on the ListView for its selection so there ListView probably gets notified first and the event then bubbles up to the TreeView/TreeViewItem.

Reply:
Hmm the fact that we have 2 separate mechanisms in place to track selection does make this a bit problematic.

The following might sound a bit crazy, but maybe switching to ItemsRepeater instead of ListView might be the better way to go here. If we switch to ItemsRepeater, we could use the ItemsRepeater SelectionModel. 

If bringing the TreeView onto the ItemsRepeater is a plan for the  future, it may be even not worth it to fix this issue, which seems problematic because of the ListView, as we would later levarage the ItemsRepeater SelectionModel anyway.

Otherwise, I think we probably would need to have different "modes" as we sometimes rely heavily on the ListVIew and in other times, we are the ones managing the SelectionModel.

@RanjeshJ What are your thoughts on this?

Reply:
Yes. This has come up before and we have had discussions about moving TreeView to ItemsRepeater but that is a bigger effort and we don't have immediate plans to do it unless there is a pressing reason to do it. Please feel free to open a proposal. In the mean time unfortunately, we will need to track selection in two different ways and bubble it up so that the TreeView control user does not see this internal detail.

Reply:
The switch to ItemsRepeater will very likely happen and should be used to fix a few issues currently introduced by the underlying ListView. That said...it might take considerable time until this fundamental change will happen and we should strive to ensure TreeView is a control ready to be used in production apps.  That means fixing bugs and providing better API when necessary (just as the the new Selection API was introduced),

Reply:
I'm now using a specific approach to track changes in multi select mode. Please have a look and drop new feedback if needed :-)